File: julia/test/timing_benchmark.jl
using LinearAlgebra, FFTW
using Test

using fsid: 
  fresp,
  estimate_cd,
  estimate_bd,
  transpose_ffdata,
  fresp,
  ffdata2fddata,
  fdestim_bd,
  fdestim_cd,
    fdsid,
    gfdsid,
  ffsid,
  bilinear_c2d,
  bilinear_d2c,
  lsim,
  fdsim,
  cf2df,
  df2cf,
  ltifr
  
# Helper functions
buildss(n, m, p) = randn(n, n), randn(n, m), randn(p, n), randn(p, m)
buildfset(N) = (0:N-1)/N
ztrans(fset) = exp.(2pi * im * fset)
computerr(fd, fde) = norm(fd-fde)/norm(fd)

function timing_fdsid()
    Nset = [100,200,400,800]
    nset = [2,5,10,20]
    m = 2
    p = 8
    MC = 100
    tim_res = zeros(length(Nset),length(nset));
    for Nidx in 1:length(Nset)
        for nidx in 1:length(nset)
            for mc in 1:MC
                N = Nset[Nidx]
                fset = buildfset(N)
                n = nset[nidx]
                A, B, C, D = buildss(n, m, p)
                lam, = eigen(A)
                rho = maximum(abs.(lam)) 
    
                ## Here we create a random stable DT system
                A = A/rho/1.01
                w = 2pi * fset
                z = ztrans(fset)
                fd = fresp(z, A, B, C, D)

                u = randn(N, m)
                y, x = lsim((A, B, C, D), u; type = Real)

                yf = fft(y, 1)
                uf = fft(u, 1)
                fddata = (z, yf, uf)
                tim_res[Nidx,nidx] = tim_res[Nidx,nidx] + @elapsed Ae, Be, Ce, De, xt, s = gfdsid(fddata, n, 2n; estTrans = true, type = Real)
                #fde = fresp(z, Ae, Be, Ce, De)
                #print( computerr(fd, fde))
            end
        end
    end
    return tim_res/MC 
end


# Main test function

print(timing_fdsid())

File: python/timing_benchmark.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Sun Jun 17 11:10:38 2018

@author: mckelvey
"""
import numpy as np
from scipy import linalg
import fsid 
import time

#import matplotlib.pyplot as plt

Nset = [100, 200, 400, 800]
nset= [2, 5, 10, 20]
m = 2;
p = 8;
MC = 100;

tim_res = np.zeros((4,4))
nidx = 0;
for n in nset:
    Nidx = 0
    for N in Nset:
        A = np.random.randn(n, n)
        lam = linalg.eig(A)[0]
        rho = np.max( np.abs(lam)) 
        ## Here we create a random stable DT system
        A = A/rho/1.01
        B = np.random.randn(n, m)
        C = np.random.randn(p, n)
        D = np.random.randn(p, m)
        fset = np.arange(0, N, dtype='float')/N
        u = np.random.randn(N, m)
        y = fsid.lsim((A, B, C, D), u, dtype='float')
    
        yf = np.fft.fft(y,axis=0)
        uf = np.fft.fft(u,axis=0)
        fset = np.arange(0, N, dtype='float')/N
        wexp = np.exp(1j*2 * np.pi * fset)
        W = np.zeros((N, p, p))
        for i in range(N):
            W[i,:,:] = np.eye(p)  
        tic =  time.mktime(time.gmtime())
        for mc in range(MC):
#            Ae, Be, Ce, De, xt, s =  fsid.gfdsid((wexp, yf, uf), n, n+1, estTrans=True, estimd=True, w=W)
             Ae, Be, Ce, De, xt, s =  fsid.gfdsid((wexp, yf, uf), n, n+1, estTrans=True, estimd=True)
        tim_res[Nidx,nidx] = time.mktime(time.gmtime())-tic
        fde = fsid.fresp(wexp, Ae, Be, Ce, De)
        fd = fsid.fresp(wexp, A, B, C, D)        
        print('|| H-He ||/||H|| = ', linalg.norm(fd-fde)/linalg.norm(fd))
        Nidx = Nidx + 1
    nidx = nidx + 1
tim_res = tim_res/MC; #timeit.timeit() has unit 1000 seconds

print tim_res
